### DISCUSSION TITLE
Problem with model class parameters initialization ?

### DISCUSSION THREAD

**Discussion by TChauvire** [JeschkeLab/DeerLab#465] [Q&A] [answered]
Hello, I'm trying to use DeerLab for analyzing data in a fully automated way. And I probably need some help to use, especially for defining Model <class> with my own parameters...
I had issue with the model definitions (and how to specify the variables).

By instance if I want to change a parameter in a model with the **"set" keyword** (Pmodel.mean.set(lb=2, ub=7, par0=5), I will get the following error message : 
    File C:\ProgramData\miniconda3\Lib\site-packages\deerlab\model.py:830 in __call__
    paramlist = self._parameter_list(order='vector')
  
    File C:\ProgramData\miniconda3\Lib\site-packages\deerlab\model.py:450 in _parameter_list
    keylist = self._vecsort(keylist)

    File C:\ProgramData\miniconda3\Lib\site-packages\deerlab\model.py:481 in _vecsort
    indices = np.concatenate([np.atleast_1d(getattr(self,param).idx) for param in dir(self) if isinstance(getattr(self,param),Parameter)])

    **ValueError: need at least one array to concatenate**

In the same way, if I want to define a model by the command : Pmodel = dl.dd_gauss(r,mean=2,std=0.1), or even Pmodel = dl.dd_gauss(r) or Pmodel = dl.dd_gauss(mean=2), I got the same type of error.

By other instance, calling Modeltest = dl.bg_hom3d(t, conc, lam) works when calling Modeltest = dl.bg_strexp(t, decay, stretch) doesn't... (I check your CheatSheet.pdf) But Modeltest.t = t, or Modeltest.decay = decay, will work...

In your examples files, there is few script using this commands, but ex_global_twostates_parametric.py is one of them.
Do you get the same error message than me ?

Best

Timothee Chauvire

**Comment by HKaras** [comment]
Hi Timothee, 

I am sorry you are having issues. This is a strange error. I use this function daily in my own automated DEER processing scripts. Please could you let me know which DeerLab and Python version you are using. 

Additionally, could you share an example script. This doesn't need to be your fully script just a short simple one where the problem occurs and I can look into it in more detail. 

Best,

Hugo

**Reply by TChauvire** [reply]
I'm using Python 3.11 with a Spyder IDE for coding.
I downloaded in September DeerLab, so it's the version 1.1.2 according to the VERSION file.

Here is a basic script where I got the error. I just to create the Model.
I will try to run it with VSCODE but I have to setup the environment first...

Best

...........................
import numpy as np
import deerlab as dl

t = np.linspace(0, 1.5, 200)
r = np.linspace(2, 7, 200)

decay1 = 1
stretch1 = 3

mean1 = 2
std1 = 0.2

Pmodeltest = dl.dd_gauss
Bmodeltest = dl.bg_strexp
Bmodeltest1 = dl.bg_hom3d(t, conc=400, lam=20)  # this works

# Error is starting appearing here...
Pmodeltest2 = dl.dd_gauss(r, mean=mean1, std=std1)
Bmodeltest2 = dl.bg_strexp(t, decay=decay1, stretch=stretch1)

**Reply by HKaras** [reply]
Hmm, This is very strange. I am having no issues runnning this code, and have tested it on multiple versions of deerlab and python. 

This was my exact test script
[test_script.py.zip](https://github.com/JeschkeLab/DeerLab/files/13161993/test_script.py.zip)

Please could you send me your full enviroment infomation. I.e. the output of `python -m pip list`

**Reply by TChauvire** [reply, marked as answer]
Ok probably my fault as I opened two instances of Spyder in the same time.
I restart spyder and it works fine this morning. I feel stupid, sorry for this. (Maybe we can close completely this discussion)

One last question, how do you define precisely the starting time ? (I did it empirically)
I was having issue to associate the different time definition from Bruker (d0, d1, d2) and the "zerotime" in DeerLab.
If you have more detailed explanations for this I'm interested...
When you simulate the background do you allow time shifting for a better simulation ?

**Reply by HKaras** [reply]
I am glad it now works. 

This figure form EFEPR workshop explains in more detail the origin of the tmin value. We are trying to move away from zerotime and deadtime as we feel they introduce confusion. 
![4pdeer_sequence_time_axes](https://github.com/JeschkeLab/DeerLab/assets/21962092/4c4d1a58-5c97-45cb-b0a4-aba25f09c27f)

The dipolar model normally allows some flexibility in the refocusing time. This is normally +- 3 times the pulse length which is 16ns by default and can be set in the experiment infomation class. If you are using chirped pulses we recomened lengthing this. You can also reparametrise the reftimes to be either delays or shifts. This is done in `dl.dipolarmodel`

> **parametrizationstring, optional**
> Parametrization strategy of the dipolar pathway refocusing times. Must be one of the following:
>
>           'reftimes' - Each refocusing time is represented individually as a parameter.
>
 >           'delays' - The pulse delays are introduced as parameters from which the refocusing times are computed. Requires experiment to be specified.
>
>           'shift' - A time shift is introduced as a parameter to represent the variability of the refocusing times from their theoretical values. Requires experiment to be specified.

**Reply by TChauvire** [reply]
Ok thank you a lot for your help.
This figure would be helpful to explain to students what the time definition they have to choose. (So here mainly tau1 as the reftime, and that works with my fitting) ;)

Have a great day.
